### PR TITLE
MEN-5216: Local HTTP proxy for Mender server API calls

### PR DESCRIPTION
**Pull request by lluiscampos**
Changelog: Mender starts a local HTTP server that will proxy incoming
requests to `/api/devices` to the currently authenticated Mender server.
The existing D-Bus API endpoints GetJwtToken and JwtTokenStateChange
will now return the local address together with the JWT token. Supports
also websocket upgrade when calling
`/api/devices/v1/deviceconnect/connect` endpoint

### CODE REVIEW

**Comment by lluiscampos** [comment]
These go together:
* https://github.com/mendersoftware/mender/pull/898
* https://github.com/mendersoftware/monitor-client/pull/100
* https://github.com/mendersoftware/integration/pull/1777

**Comment by lluiscampos** [comment]
@mender-test-bot start pipeline --pr monitor-client/pull/100 --pr integration/pull/1777

**Comment by lluiscampos** [comment]
@mender-test-bot start pipeline --pr monitor-client/100 --pr integration/1777

**Comment by merlin-northern** [comment]
I am leaving this one to Kristian, as he already took a look.

File: app/auth.go [hunk 1]
```diff
@@ -527,6 +545,14 @@ func (m *menderAuthManagerService) broadcast(message AuthManagerResponse) {
 // broadcastAuthTokenStateChange broadcasts the notification to all the subscribers
 func (m *menderAuthManagerService) broadcastAuthTokenStateChange() {
 	authToken, err := m.authToken()
+
+	// reconfigure proxy
+	m.localProxy.Stop()
+	if authToken != "" {
+		_ = m.localProxy.Reconfigure(m.serverURL, string(authToken))
+		m.localProxy.Start()
```
[the remark this answers is not in the document]
Comment (lluiscampos) [reply app/auth.go:553]: No, it is not correct! I'll fix. It is not a lot we can do at this point, but a good error message will help diagnostics.

Side note: my original code was just plain ignore, but after the static checks it forced me to introduce `_ = ...` without much thinking but probably thanks to that you caught the obvious issue! Good for the linter.

File: app/proxy/proxy.go [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+	if pc.conf.listener != nil {
+		pc.isRunning = true
+
+		initDone := make(chan struct{}, 1)
+		go pc.run(initDone)
+
+		// Wait for the server to start
+		<-initDone
+
+		runtime.SetFinalizer(pc, func(pc *ProxyController) {
```
[the remark this answers is not in the document]
Comment (lluiscampos) [reply app/proxy/proxy.go:195]: I've seen this inner struct mechanism but didn't fully understand it. I am looking into it and adding such design here. Thanks!

File: app/proxy/proxy_test.go [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+	srv := cltest.NewClientTestServer()
+	defer srv.Close()
+
+	srv.Update.Has = false
+	proxyController, err := NewProxyController(&http.Client{}, nil, srv.Server.URL, "BringItOn")
+	require.NoError(t, err)
+
+	for i := 0; i < 1000; i++ {
+		stopReconfigureStart(t, proxyController, srv.Server.URL, "BringItOn")
+		req, err := newDeploymentsNextRequest(proxyController)
```
[the remark this answers is not in the document]
Comment (lluiscampos) [reply app/proxy/proxy_test.go:216]: Heheh. After few spurious race conditions this was the only way I felt confident I would fix it for good!

File: app/proxy/proxy.go [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+	pc.isRunning = false
+
+	runtime.SetFinalizer(pc, nil)
+}
+
+func (pc *ProxyController) run(initDone chan struct{}) {
+	mux := http.NewServeMux()
+	mux.HandleFunc(ApiUrlDevicesPrefix, pc.checkAuthorizationHook(pc.doHttpRequest))
+	mux.HandleFunc(ApiUrlDevicesAuthentication, pc.apiDevicesAuthenticationHandler)
+	mux.HandleFunc(ApiUrlDevicesConnect, pc.checkAuthorizationHook(pc.apiDevicesConnectHandler))
```
[the remark this answers is not in the document]
Comment (lluiscampos) [reply app/proxy/proxy.go:223]: websockets protocol does not use `CONNECT` method. It works with a `GET` that includes headers `Connection: Upgrade` and `Upgrade: websocket`, to which the server responds with code `101 Switching Protocols` and both parties agreed then to keep the TCP connection alive.

Mozilla [docs](https://developer.mozilla.org/en-US/docs/Web/HTTP/Protocol_upgrade_mechanism) are simple but pretty good.

I haven't read the RFC in detail, but I think there is no way to distinguish beforehand because even this `ws://` or `wss://` is a convenience thing, you can actually use `http://` and `https://` and include such headers.

In short, either we have a hook that checks for this headers and then passes to one handler or the other, or we leave it as is and add individual handlers for individual endpoints as we go...

File: app/proxy/proxy_ws.go [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+)
+
+const (
+	maxWsConnections = 1
+)
+
+var (
+	// From https://github.com/mendersoftware/deviceconnect/blob/master/api/http/device.go
+	upgrader = websocket.Upgrader{
+		Subprotocols: []string{"protomsg/msgpack"},
```
[the remark this answers is not in the document]
Comment (lluiscampos) [reply app/proxy/proxy_ws.go:35]: This is the way the _server_ specifies which websocket subprotocols it supports. It cannot be fetched from the incoming headers because then you would be "supporting" whatever the client asks for. In Mender backend we only support `protomsg/msgpack` ATM.

File: app/proxy/proxy_ws_test.go [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+	require.Equal(t, http.StatusSwitchingProtocols, resp.StatusCode)
+	defer conn.Close()
+
+	// bi-directional connections established
+	goRoutinesConnOpen := runtime.NumGoroutine()
+	assert.Greater(t, goRoutinesConnOpen, goRoutinesStart)
+
+	mClose := websocket.FormatCloseMessage(websocket.CloseNormalClosure, "done for the day")
+	conn.WriteMessage(websocket.CloseMessage, mClose)
+	time.Sleep(1 * time.Millisecond)
```
[the remark this answers is not in the document]
Comment (lluiscampos) [reply app/proxy/proxy_ws_test.go:251]: I didn't think about `assert.Eventually` :sweat_smile:  I'll rework all these waits, thanks!

[the remark this answers is not in the document]
Comment (lluiscampos) [reply app/proxy/proxy_ws_test.go:251]: Wait, on a second thought this `TestProxyWsGoroutines` is a bit special.

I got inspiration from [this test](https://github.com/mendersoftware/mender/blob/master/app/auth_test.go#L634), and the idea is to "wait a bit" in the different steps for the goroutines to start or GC to kill them. With an `assert.Eventually` step by step it won't work the same way... I could way wait until num goroutines increases, but that would exit on >1 and maybe there are more to be started...

The test was very useful while developing, maybe not so much now. If we are concern about the timing I can do something like:
* measure num routines
* start
* wait for routines to increase
* do everything else (establish websocket, send data, etc)
* wait for routines to be the same as in the beginning.

File: app/proxy/proxy_ws_test.go [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+		t.FailNow()
+	}
+
+	assert.True(t, reflect.DeepEqual(srv.Connect.SendMessages, recvMessages))
+	assert.True(t, reflect.DeepEqual(sendMessages, srv.Connect.RecvMessages))
+
+	mClose := websocket.FormatCloseMessage(websocket.CloseNormalClosure, "done for the day")
+	conn.WriteMessage(websocket.CloseMessage, mClose)
+	time.Sleep(1 * time.Millisecond)
+	assert.Equal(t, websocket.CloseNormalClosure, srv.Connect.RecvClose)
```
[the remark this answers is not in the document]
Comment (lluiscampos) [reply app/proxy/proxy_ws_test.go:147]: Yes, this one I fixed now with `assert.Eventually` :+1:

File: app/auth.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (kacf) [reply app/auth.go]: Indeed, good linter!

File: app/proxy/proxy.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (kacf) [reply app/proxy/proxy.go:233]: You're right, I confused websockets and CONNECT. Never mind then.

File: app/proxy/proxy_ws.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (kacf) [reply app/proxy/proxy_ws.go:35]: Discussed on Slack, will look into it, but outcome is still uncertain.

File: app/proxy/proxy_ws_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (kacf) [reply app/proxy/proxy_ws_test.go:255]: Yes, that sounds more robust!

File: app/proxy/proxy_ws.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (lluiscampos) [reply app/proxy/proxy_ws.go:35]: Right, so what I have implemented is for the proxy:
* on client to backed direction, propagate `Sec-WebSocket-Protocol` header from the client (handshake request)
* on backend to client direction, support only `Sec-WebSocket-Protocol` header from the backend (handshake response)


```
CLIENT -----------------------> PROXY -----------------------> BACKEND
       protocol1, protocol 2          protocol1, protocol 2

CLIENT <----------------------- PROXY <----------------------- BACKEND
                      protocol1                      protocol1
```

I added one unit test and made also some local experiments. It should be fine

[the remark this answers is not in the document]
Comment (kacf) [reply app/proxy/proxy_ws.go]: Good, that's exactly what I meant, and what I hoped it would support!

File: app/proxy/proxy_ws_test.go [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+	// assert.Eventually creates one Goroutine on its own, so take that into account
+	// time.Sleep(100 * time.Millisecond)
+	assert.Eventually(
+		t,
+		func() bool {
+
+			return runtime.NumGoroutine() == goRoutinesStart+1
+		},
+		100*time.Millisecond,
+		1*time.Millisecond,
```
Comment (kacf) [inline app/proxy/proxy_ws_test.go:309]: 👍

File: client/test/server.go [hunk 9]
```diff
@@ -212,6 +243,30 @@ func (cts *ClientTestServer) authReq(w http.ResponseWriter, r *http.Request) {
...
+		for hdr := range cts.RequestHeader.Header {
+			h := r.Header.Get(hdr)
+			if h == "" {
+				log.Errorf("header %s not found, got %+v, expected %+v",
+					hdr, r.Header, cts.RequestHeader.Header)
+				w.WriteHeader(http.StatusBadRequest)
+				return
+			}
+			if h != cts.RequestHeader.Header.Get(hdr) {
+				log.Errorf("header %s missmatch, got %+v, expected %+v",
```
Comment (kacf) [inline client/test/server.go:257]: Nitpick: `s/missmatch/mismatch/`

**Review by kacf — APPROVED** [review]

File: app/proxy/proxy_ws.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (lluiscampos) [reply app/proxy/proxy_ws.go]: Yes! Thanks again for voicing this concern, I think we got it right now :tada:

**Comment by lluiscampos** [comment]
Extended the D-Bus acceptance tests to verify that the Server URL returned by the API is actually `localhost`. See:
* https://github.com/mendersoftware/meta-mender/pull/1537

**Comment by lluiscampos** [comment]
@mender-test-bot start pipeline --pr monitor-client/100 --pr integration/1777 --pr meta-mender/1537 :crossed_fingers:
